### PR TITLE
Add createPipelineRun to Pipeline Controller

### PR DESCRIPTION
**Pull request by tejal29**
This addresses a part of #61.
In this commit, we are adding createPipelineRun to PipelineRun Controller.
This method will be used in PipelineRun.reconcile in future when we detect the pipelineRun
 does not have tasksruns for tasks in the pipeline.

### CODE REVIEW

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -144,21 +150,74 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
-	if _, err := c.pipelineLister.Pipelines(pr.Namespace).Get(name); err != nil {
-		c.Logger.Errorf("%q failed to Get Pipeline: %q",
-			fmt.Sprintf("%s/%s", pr.Namespace, pr.Name),
-			fmt.Sprintf("%s/%s", pr.Namespace, name))
-		return nil
+	if _, err := c.getPipelineIfExists(pr); err != nil {
+		return err
 	}
-
 	// TODO fetch the taskruns status for this pipeline run.
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: should probably call createPipelineRun() here

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -144,21 +150,74 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
 
 	// TODO check status of tasks and update status of PipelineRuns
 
 	return nil
 }
 
+func (c *Reconciler) createPipelineRun(ctx context.Context, p *v1alpha1.Pipeline) error {
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: this func is actually not creating the pipelineRun, its creating the contents of the pipelineRun, so maybe we can add something to the name to make it more clear. Maybe `createPipelineRunTaskRuns` or something?

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 3]
```diff
@@ -144,21 +150,74 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      trName,
+			Namespace: t.Namespace,
+		},
+	}
+	return tr, nil
+}
+
+func randomizeTaskRunName(t *v1alpha1.Task) string {
+	return util.AppendRandomString(t.Name)
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: can we just use the name of the pipeline in combination with the name of the task?

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 4]
```diff
@@ -144,21 +150,74 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
+}
+
+func (c *Reconciler) getPipelineIfExists(pr *v1alpha1.PipelineRun) (*v1alpha1.Pipeline, error) {
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: no need for `IfExists` in the name

File: pkg/util/randstring.go [hunk 5]
```diff
@@ -11,7 +11,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package test
+package util
```
Comment (pivotal-nader-ziada) [inline pkg/util/randstring.go]: I don't think we need this

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 6]
```diff
@@ -144,21 +150,74 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
 
 	return nil
 }
 
+func (c *Reconciler) createPipelineRun(ctx context.Context, p *v1alpha1.Pipeline) error {
+	tasks, err := c.getTasks(p)
+	if err != nil {
+		return errors.NewBadRequest(fmt.Sprintf("pipeline %q is not valid due to %v", p.Name, err))
+	}
+	for _, t := range tasks {
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: check my comment in the review, should we create all taskRun or just the first one and taskRun controller can do the rest?

**Comment by tejal29** [comment]
@pivotal-nader-ziada Thanks! I completely forgot the flow.  For the initial demo we were targeting linear pipeline. However, in future, it should be able to run tasks in parallel. we probably need to have a method  that should check if all inputs for a given task are available and then run.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 3 above)
Comment (tejal29) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: i can add pipeline-run name in combination of name of task hoping we would a unique name for each pipeline-run.

**Comment by pivotal-nader-ziada** [comment]
looks good, but has conflicts now because of #97 

Just one nit: can you remove the `fmt.Println(p)` in `pipelinerun.go`

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 7]
```diff
@@ -144,21 +154,86 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
+	}
+	for _, pt := range p.Spec.Tasks {
+		trName := getTaskRunName(prName, &pt)
+		_, listErr := c.taskRunLister.TaskRuns(p.Namespace).Get(trName)
+		if errors.IsNotFound(listErr) && c.canTaskRun(&pt) {
+			tr, err := createTaskRun(taskRefMap[pt.Name], trName)
+			if err != nil {
+				return err
+			}
+			c.PipelineClientSet.PipelineV1alpha1().TaskRuns(p.Namespace).Create(tr)
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: Can we move the this line inside the createTask func, also check for any errors coming back

**Comment by tejal29** [comment]
@pivotal-nader-ziada This is ready for last look.  Sorry, i messed the commit history. Instead of running `git commit -m` i ran `git commit --amend`

**Pull request merged**